msdk-io/msdk-io-mzml2/src/main/java/io/github/msdk/io/mzml2/MzMLSpectrum.java

```diff
@@ -154,95 +148,153 @@ public Integer getNumberOfDataPoints() {
 
       compressions.add(getIntensityBinaryDataInfo().getCompressionType());
 
-      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream, 0,
+      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream,
+          getIntensityBinaryDataInfo().getPosition(),
           getIntensityBinaryDataInfo().getEncodedLength());
       InputStream decodedIs = Base64.getDecoder().wrap(encodedIs);
       byte[] decodedData = IOUtils.toByteArray(decodedIs);
+      System.out.println(new String(IOUtils.toByteArray(encodedIs)));
 
-      result = MzMLIntensityPeaksDecoder.decode(decodedData, 0, precision,
+      intensityValues = MzMLIntensityPeaksDecoder.decode(decodedData, decodedData.length, precision,
           getIntensityBinaryDataInfo().getArrayLength(), compressions).arr;
     } catch (Exception e) {
       throw (new MSDKRuntimeException(e));
     }
 
-    return result;
+    return intensityValues;
   }
 
   @Override
-  public Float getTIC() {
-    // TODO Auto-generated method stub
-    return null;
+  public MsSpectrumType getSpectrumType() {
+    MsSpectrumType spectrumType = SpectrumTypeDetectionAlgorithm.detectSpectrumType(mzValues,
+        intensityValues, getMzBinaryDataInfo().getArrayLength());
+    return spectrumType;
   }
 
   @Override
-  public Range<Double> getMzRange() {
-    // TODO Auto-generated method stub
-    return null;
+  public Float getTIC() {
+    Float tic = MsSpectrumUtil.getTIC(intensityValues, getMzBinaryDataInfo().getArrayLength());
+    return tic;
   }
 
   @Override
-  public MzTolerance getMzTolerance() {
-    // TODO Auto-generated method stub
-    return null;
+  public Range<Double> getMzRange() {
+    Range<Double> mzRange =
+        MsSpectrumUtil.getMzRange(mzValues, getMzBinaryDataInfo().getArrayLength());
+    return mzRange;
   }
 
   @Override
   public RawDataFile getRawDataFile() {
-    // TODO Auto-generated method stub
-    return null;
+    return dataFile;
   }
 
   @Override
   public Integer getScanNumber() {
-    // TODO Auto-generated method stub
-    return null;
+    return scanNumber;
   }
 
   @Override
   public String getScanDefinition() {
-    // TODO Auto-generated method stub
-    return null;
+    return getCVValue(MzMLCV.cvScanFilterString);
   }
 
   @Override
   public MsFunction getMsFunction() {
-    // TODO Auto-generated method stub
-    return null;
+    Integer msLevel = 1;
+    String value = getCVValue(MzMLCV.cvMSLevel);
+    if (!Strings.isNullOrEmpty(value))
+      msLevel = Integer.parseInt(value);
+    return MSDKObjectBuilder.getMsFunction(msLevel);
   }
 
   @Override
   public MsScanType getMsScanType() {
-    // TODO Auto-generated method stub
-    return null;
+    return MsScanType.UNKNOWN;
   }
 
   @Override
   public Range<Double> getScanningRange() {
```

Comment:
This value can also be parsed from the mzML:
````
 <scanList count="1">
            <cvParam cvRef="MS" accession="MS:1000795" name="no combination" value=""/>
            <scan>
              <cvParam cvRef="MS" accession="MS:1000016" name="scan start time" value="0.0028779136" unitCvRef="UO" unitAccession="UO:0000031" unitName="minute"/>
              <cvParam cvRef="MS" accession="MS:1000512" name="filter string" value="FTMS + p ESI Full ms [118.00-1750.00]"/>
              <cvParam cvRef="MS" accession="MS:1000616" name="preset scan configuration" value="1"/>
              <cvParam cvRef="MS" accession="MS:1000927" name="ion injection time" value="50.000000745058" unitCvRef="UO" unitAccession="UO:0000028" unitName="millisecond"/>
              <scanWindowList count="1">
                <scanWindow>
                  <cvParam cvRef="MS" accession="MS:1000501" name="scan window lower limit" value="118.0" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
                  <cvParam cvRef="MS" accession="MS:1000500" name="scan window upper limit" value="1750.0" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
                </scanWindow>
              </scanWindowList>
            </scan>
          </scanList>
````
Check the "scan window lower limit" and "scan window upper limit" accession. If the value is not present in the file, return the value of getMzRange()